railties/lib/rails/commands/server/server_command.rb

```diff
@@ -10,6 +10,33 @@
 
 module Rails
   class Server < ::Rack::Server
+    module HandlerSuggestion
+      include Rails::Command::Behavior
+
+      # Hard-coding a bunch of handlers here as we don't have a public way of
+      # querying them from the Rack::Handler registry.
+      HANDLERS = %w(cgi fastcgi webrick lsws scgi thin puma unicorn)
+
+      def self.message(handler)
+        if handler.in?(HANDLERS)
+          <<~MSG
+            Could not load server "#{handler}". Maybe you need the add it to the Gemfile?
+
+              gem "#{handler}"
+
+            Run `rails server --help` for more options.
+          MSG
+        else
+          suggestions = HANDLERS.sort_by { |h| levenshtein_distance(handler, h) }.map(&:inspect)
+
+          <<~MSG
+            Could not find server "#{handler}". Maybe you meant #{suggestions.first} or #{suggestions.second}?
+            Run `rails server --help` for more options.
+          MSG
+        end
```

Comment:
Doing argv validation seems like it should a concern of the command and not the server itself. Both `include Rails::Command::Behavior` and `rails server --help` seem to allude as much.

Could we move it to the command if we do this?

```ruby
capturing_missing_handler do
  server.start
end
```

Reply:
Tried that as well – it's still messy because `levenshtein_distance` is a private class method:

```ruby
        def suggest_rack_server(handler)
          if handler.in?(HANDLERS)
            <<~MSG
              Could not load server "#{handler}". Maybe you need the add it to the Gemfile?

                gem "#{handler}"

              Run `rails server --help` for more options.
            MSG
          else
            suggestions = HANDLERS.sort_by { |h| self.class.send(:levenshtein_distance, handler, h) }.map(&:inspect)

            <<~MSG
              Could not find server "#{handler}". Maybe you meant #{suggestions.first} or #{suggestions.second}?
              Run `rails server --help` for more options.
            MSG
          end
        end
```